tt_metal/tt_stl/overloaded.hpp

```diff
@@ -0,0 +1,38 @@
+// SPDX-FileCopyrightText: © 2024 Tenstorrent Inc.
+//
+// SPDX-License-Identifier: Apache-2.0
+
+#pragma once
+
+namespace tt::stl {
+
+// `overloaded` allows to combine multiple lambdas into a single object with the overloaded `operator()`.
+// This is useful for creating visitor objects for using in `std::visit`, for example:
+//
+// void f(const std::variant<int, std::string>& v) {
+//   std::visit(overloaded {
+//     [](int i) { std::cout << "Got an int: " << i << std::endl; },
+//     [](const std::string& str) { std::cout << "Got a string: " << str << std::endl; },
+//   }, v);
+// }
+//
+//
+// Generic lambdas can also be used to simplify handling of variant types:
+//
+// void f(const std::variant<std::vector<int>, int, std::string>& v) {
+//   std::visit(overloaded {
+//     [](const std::vector<int>& vec) { std::cout << "Got a vector of size: " << vec.size() << std::endl; },
+//     [](const auto& other) { std::cout << "Got something else: " << other << std::endl; },
+//   }, v);
+// }
```

Comment:
Again, my concern is that visually this appears to be advocated over simply
```cpp
void f(int i) {
  // Do something
}

void f(std::string_view str) {
  // Do something
}
```
whereas the point of `overloaded` as opposed to an overload set is to reuse other boilerplate that might exist in the function like
```cpp
  std::for_each(v.begin(), v.end(), ...);
```
I don't see it as "noise", but rather showing that this code is implicitly re-used rather than duplicated when using `overloaded` instead of an overload set.

Reply:
Discussed offline; we decided to move to an example that relies on a declaration only, for simplicity.